CupertinoSheetRoute

MitchellGoodwin opened the pull request:
Fixes #42560

### Updated 12/04/24

Adds a CupertinoSheetRoute which displays an iOS style sheet, while giving the previous route a delegated transition to sync up correctly.

There's a lot of sample files in this draft showing different use cases for the sake of reviewing the PR. We'll want to probably go down to about two before this PR lands.

**TODO in this PR**

- [x] Write actual documentation
- [x] Write tests

**TODO in a feature branch, and will be merged with this PR**

- [x] Add swipe down to dismiss

**TODO in a follow up PR**

- [ ] Stretching on swipe up
- [ ] Add a fullscreen version
- [x] Dark mode theming

How it currently compares, with the simple case and nested navigation:
| Flutter  | Native|
| ------------- | ------------- |
| ![Flutter-Simple](https://github.com/user-attachments/assets/2a1f277e-91c9-48e0-b894-5fad71ef6a21) |![Native-Simple](https://github.com/user-attachments/assets/6960252a-f762-4ad8-8d68-5c3d7cf8d4e7) |
| ![Flutter-Nested](https://github.com/user-attachments/assets/bacb3e35-9d11-4113-9331-75daaded67e7) | ![Native-stacked](https://github.com/user-attachments/assets/70ccbbbb-24c2-40b5-b838-4f8412828b9b) |

Update:
![Flutter-Status-Bar](https://github.com/user-attachments/assets/c738cc5c-7176-4df2-9422-6b3fa608c943)
Now the status bar will transition from dark to light text so that it's visible after the transition.

`cupertino_sheet.0.dart` has an example of what it looks like to do a simple case. It's fairly straightforward.

`cupertino_sheet.1.dart` shows one with nested navigation. It's necessary to add a `Navigator` if showing page navigation within the sheet is wanted. I'm thinking adding an API option that auto sets this up may be needed as there are some pitfalls.

`cupertino_sheet.2.dart` functionally does the same as above but uses the `showCupertinoSheet` to reduce boilerplate.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

MitchellGoodwin committed:
Start sheet widget

MitchellGoodwin committed:
Add inherited controller

MitchellGoodwin committed:
add examples

MitchellGoodwin committed:
Adjust transition

MitchellGoodwin commented on examples/api/lib/cupertino/sheet/cupertino_sheet.1.dart:
```diff
@@ -0,0 +1,173 @@
+            ),
+            CupertinoButton.filled(
+              onPressed: () {
+                Navigator.of(topLevelContext ?? context).push(CupertinoSheetRoute<void>(
+                  pageBuilder: (BuildContext context) => const SheetContentWithNavigator()
```
Because of how odd this is, rather than having people use the `CupertinoSheetRoute`, it's probably better to do something like `showCupertinoSheet`.

justinmc replied:
I mentioned it elsewhere but I agree that the showCupertinoSheet is probably worth it for convenience. Though what's your concern with the complexity here, is it the need to use CupertinoSheetRoute and to get the BuildContext right? Compare that with a normal usage of `Navigator.of(context).push` like you have below:

```dart
Navigator.of(context).push(
  CupertinoPageRoute<void>(builder: (BuildContext context) => const SheetNextPage())
);
```

MaherSafadii commented:
I just tried it locally and it matches the native one animation wise, but on the native one when you drag upwards, it stretches a bit, I hope you can also implement this behaviour too.

MitchellGoodwin commented:
> I just tried it locally and it matches the native one animation wise, but on the native one when you drag upwards, it stretches a bit, I hope you can also implement this behaviour too.

Thanks for taking a look! Good callout on the drag upwards stretch. We're going to probably add the drag controls in a feature branch off of this one. We might not focus on getting the drag upwards behavior correct for initially landing it in master, and instead have that in a follow up issue for the sake of iteration. Same with the fullscreen version. I'll add it to the todos though so it's tracked somewhere.

MaherSafadii commented:
another notice, the status bar should automatically turn the brightness to light when the sheet is opened.

hls-app commented:
This is awesome! Been using https://github.com/jamesblasco/modal_bottom_sheet and would definitely prefer a native solution!

MitchellGoodwin commented:
> another notice, the status bar should automatically turn the brightness to light when the sheet is opened.

Thanks for the callout. We can change the status bar brightness when the sheet is opened, but it is sudden and we can't currently get the gradual transition native has with our APIs. I talked to the iOS team about that and they are aware of it. I'll open an iterative issue on that when this lands.

MitchellGoodwin committed:
Add showCupertinoSheet

MaherSafadii commented:
Another thing to consider is that is in dark mode the sheet should light as the opposite of what happens in light mode so the previous sheet becomes visible if its black: 

https://github.com/user-attachments/assets/0e20ce8d-526d-4d25-99f0-f921ff32863c

MitchellGoodwin commented:
> Another thing to consider is that is in dark mode the sheet in the back should lighten as the opposite of what happens in light mode it darkens, it makes the sheet in the background become visible if its black:

It looks to me like the top sheet itself has a brighter background as well so it stands out. That's probably default behavior. I'll add dark mode to the to-dos.

MaherSafadii commented:
> > Another thing to consider is that is in dark mode the sheet in the back should lighten as the opposite of what happens in light mode it darkens, it makes the sheet in the background become visible if its black:
> 
> It looks to me like the top sheet itself has a brighter background as well so it stands out. That's probably default behavior. I'll add dark mode to the to-dos.

Yeah it does, on ios it detects if a sheet is open and automatically sets the sheet in the fronts background color to secondaryBackground.

justinmc commented on examples/api/lib/cupertino/sheet/cupertino_sheet.2.dart:
```diff
@@ -0,0 +1,145 @@
+          children: <Widget>[
+            CupertinoButton.filled(
+              onPressed: () {
+                showCupertinoSheet<void>(
```
Maybe in the final PR we should include an example that shows getting a return value from this.

justinmc commented on examples/api/lib/cupertino/sheet/cupertino_sheet.2.dart:
Well this is does look a lot simpler when using showCupertinoSheet.

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,349 @@
+
+  /// The primary delegated transition. Will slide a non [CupertinoSheetRoute] page down.
+  ///
+  /// If a [CupertinoSheetController] already exists in the stack, then it will
```
> in the stack

Actually, if there is a regular route between the existing CupertinoSheetRoute and the new one, then it should it do the normal slide down animation?

MitchellGoodwin replied:
Right now if a regular route gets pushed to the root navigator after the sheet than the normal delegate logic will happen. I'm not 100% sure what the correct behavior should be. We may want to not play any outgoing transition at all on the sheet.

justinmc replied:
So it won't really be possible to push a normal route on top of a MBS and have it display as a normal route (instead it displays as an additional sheet in the MBS). Can you confirm that we're ok not supporting that? If it's not possible on native iOS or recommended against then this is probably the right call.

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,349 @@
+  required BuildContext context,
+  required WidgetBuilder pageBuilder,
+  bool useRootNavigator = true,
+  bool useNestedNavigation = false,
```
Can you try to get state restoration working here? Whether in this code or in the example code. I really wonder what this would look like in a fully-fledged production navigation scenario, in order to decide what to do here.

MitchellGoodwin replied:
Sorry, I talked about how iOS does it [outside this comment](https://github.com/flutter/flutter/pull/157568#issuecomment-2480036684). TLDR is on iOS when you add a sheet it automatically adds something like a Navigator to the content so nested navigation happens by default. If you want to do unnested navigation you have to pass a callback to the sheet content that will trigger navigation outside of the scope of the sheet.

chunhtai replied:
ah I see, I think the native ios behavior is the same as if we add navigator by default. If user want to add a unnest page they need to do Navigator .push with the context above the cupertino sheet.

I don't have strong opinion on whether we need to follow them. your call.

justinmc reviewed (commented):
Sorry it took me so long to review this. I played around with it locally and it seems to feel really good to my Android-user eyes. The API and examples also seem straightforward and well-integrated with Flutter's way of doing navigation.

### Comparison with Swift UI

In Swift UI this is done something like this right:

```swift
.sheet(isPresented: $isSheetOpen)
```

How about to add another sheet? How about to add another page to the current sheet?

I'm wondering what the main flows are there so that we can make them easy in Flutter too.

### Other stuff

Pushing a CupertinoSheetRoute starts showing the "MBS" style navigation, then:
 * Pushing another CuperinoSheetRoute appears like another tab within the MBS navigation.
 * Pushing any other route does it's normal transition, as if moving away from the MBS.
 * If you want to push something inside of the topmost MBS tab, you have to do that with nested navigation.

In Swift UI is it possible to do the second bullet point above?

Also, I think you mentioned it briefly before, but I wonder how you plan on implementing the feature where you can drag down to dismiss the sheet?

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,349 @@
+  });
+
+  /// Docs placeholder
+  final WidgetBuilder pageBuilder;
```
`pageBuilder` is used for some dialog routes. `builder` is more common though, so we'll go with that.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,349 @@
+          return CupertinoPageRoute<void>(
+            builder: (BuildContext context) {
+              return PopScope(
+                canPop: settings.name != '/',
```
Yes, If you call `pop()` programmatically I want on the initial route in the sheet, I want it to pop the sheet. Originally it was going to an inbetween state.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,349 @@
+  /// wraps all of the sheets in the stack. Usefull for adding a new
+  /// [CupertinoSheetRoute], or any page route that needs to animate outside the
+  /// bounds of the sheet.
+  final BuildContext topLevelContext;
```
Probably, actually.

MitchellGoodwin commented:
> Also, I think you mentioned it briefly before, but I wonder how you plan on implementing the feature where you can drag down to dismiss the sheet?

I'm working on that locally. There will be another draft PR branched off of this one for that.

chunhtai commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,349 @@
+}
+
+/// Docs placeholder
+class CupertinoSheetController extends InheritedWidget {
```
This needs a better name. Controller is usually a changeNotifier object in Flutter.

chunhtai replied:
instead of a inherited widget that people can access different level of context, have you thought about have a controller object that provide API to pop or push navigator route at different level?

chunhtai replied:
The reason for this is that holding a context as widget parameter is can be error prompt as it can go stale at anytime. it will be more dangerous if this is a public api

justinmc replied:
The only place that the `context` parameter is used is in popSheet, right? Maybe that could just take BuildContext as a parameter so that CupertinoSheetController doesn't need to.

justinmc replied:
> instead of a inherited widget that people can access different level of context, have you thought about have a controller object that provide API to pop or push navigator route at different level?

@chunhtai Are you talking about a generic controller to control nested navigation, or something specifically for this modal bottom sheet?

chunhtai replied:
for this modal bottom sheet. For general nested navigation once can already use NavigatorState if they possess global key

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,349 @@
+  final WidgetBuilder builder = switch (useNestedNavigation) {
+    false => pageBuilder,
+    true => (BuildContext context) {
+      return Navigator(
+        initialRoute: '/',
+        onGenerateRoute: (RouteSettings settings) {
+          return CupertinoPageRoute<void>(
+            builder: (BuildContext context) {
+              return PopScope(
```
Consider trying this out with
 * go_router
 * deep linking, and/or specific web-style URL requirements.

I wonder what kind of requirements you'll run into to get those working. Maybe you will need some kind of builder parameter instead of just the plain boolean.

kerberjg commented:
@MitchellGoodwin Great work so far! Looking forward for this to land 💖

MitchellGoodwin committed:
Refactor names and context usage

MitchellGoodwin committed:
Update sample

MitchellGoodwin committed:
Add restorable state with nested navigation sample

MitchellGoodwin commented:
Added a sample that has both state restoration and nested navigation in one file. It works locally, but I'm unsure if I set it up in the ideal way.

justinmc reviewed (commented):
Some thoughts about the nested Navigator situation below.

MitchellGoodwin commented:
> So it won't really be possible to push a normal route on top of a MBS and have it display as a normal route (instead it displays as an additional sheet in the MBS).

It will be possible. The previous route, which is the MBS, just wouldn't transition out and would be static. So a Cupertino page would slide in from the right, but the previous route would not slide left.

On native iOS, it is possible, just a little awkward. If you navigate to a new page from within the sheet, it will be nested within the sheet. To navigate directly from the sheet to new page outside of the sheet, you have to define something like a callback in the parent view and pass it down to the sheet.

I don't believe this is recommended (it's difficult to find direct guidance on this). The sheet is supposed to give options related to the context of the screen it's covering, then always go back to the original page when it's done.

> My opinion is that we should leave nested navigation up to the user completely

I think I'm at that opinion as well. I've gotten a nested navigation example with the pages API working in several different ways, none of which make for an short sample. I think it would be pretty bloated to give a FULL sample in the PR, but could make a good blogpost/ article for the website.

Again on native, I believe the way it works is the `.sheet` adds a NavigationView or NavigationStack with a lot of defaults turned off to the sheet by default. Both of which are very similar to our own Navigator. However, the recommendation seems to be that if you want to have nested navigation in your app it's better to add your own NavigationStack to the body of the sheet to control that.

So I think removing `useNestedNavigation` is the best idea. I was a thinking of providing a `navigatorBuilder` could be an option, but again, it's probably better to just let the developer handle setting that up in what ways work for them. If a common boiler plate becomes needed I'm sure people will let us know/ a package will be created.

I do not think we need or should add a Navigator by default the way SwiftUI does. That's not very "Flutter"y.

cc @chunhtai

justinmc commented:
> It will be possible. The previous route, which is the MBS, just wouldn't transition out and would be static. So a Cupertino page would slide in from the right, but the previous route would not slide left.

Can you explain how to do that in Flutter code? If that's so, and as you explain it's hard to do in Swift UI, then I think I'm on board with the current approach 👍 .

MitchellGoodwin commented:
> Can you explain how to do that in Flutter code? If that's so, and as you explain it's hard to do in Swift UI, then I think I'm on board with the current approach 👍 .

If there is no nested Navigator, than you just push a route as normal. If there *is* a nested Navigator, than you need to push the route to the root Navigator. So `Navigator.push(context, rootNavigator: true)`, or if using named routes, whatever name gets you to the root Navigator.

chunhtai commented:
> So I think removing useNestedNavigation is the best idea. I was a thinking of providing a navigatorBuilder could be an option, but again, it's probably better to just let the developer handle setting that up in what ways work for them. If a common boiler plate becomes needed I'm sure people will let us know/ a package will be created.

>I do not think we need or should add a Navigator by default the way SwiftUI does. That's not very "Flutter"y.

I have not looked at the latest change yet, so I may be under false assumption.

In general, It depends on how hard the set up will be, and usually asking developer to create Navigator will require additional work like wrapping it in PopScope or navigatorPopHandler. If it does indeed have this kind of caveats AND it is a common use case, I would suggest adding a parameter to create navigator for them. Imagine we do the opposite, the developers make a bunch of mistakes and create  apps with bad user experience. That is not something we want

justinmc commented:
> If there is no nested Navigator, than you just push a route as normal. If there is a nested Navigator, than you need to push the route to the root Navigator. So Navigator.push(context, rootNavigator: true), or if using named routes, whatever name gets you to the root Navigator.

Ok got it now. Just to explain the whole situation for posterity (and myself), there are three ways to add a new route to an MBS:

 1. Push another MBS route to the root Navigator. This will add another sheet to the tabbed view.
 2. Push a route to the nested Navigator. This will cause navigation within the MBS while keeping the tabbed view visible.
 3. Push a non-MBS route to the root Navigator. This will do a full-screen transition to a new page on top of the MBS, hiding the tabbed view.

justinmc commented:
@chunhtai I talked about this with @MitchellGoodwin and I think you're right. I think we should do the nested navigation for the user if they opt in, because most users won't care about things like setting up their own deep linking for what should be an ephemeral dialog. Those that do can pass their own Navigator without opting in.

I'm assuming that:

 * Most users just want a single MBS sheet, which does not require nested navigation.
 * The next most users will want multiple MBS sheets, which also does not require nested nav.
 * A minority of users will want navigation within an MBS sheet, requiring nested nav. Since this is a less common use case, users have to either opt in by setting or `useNestedNavigation: true`, or pass their own Navigator.
   - Among these users, most fall into the category of wanting nested navigation done for them. They don't care about advanced navigation features like deep linking, because the MBS is just an ephemeral popup sort of thing.
   - The rest of these users are advanced and want to build their own Navigator, and they are welcome to do that by not setting `useNestedNavigation`.

That way we optimize for what I think will be the most common use cases while still keeping advanced uses possible.

MitchellGoodwin committed:
Improve documentation, add tests

MitchellGoodwin committed:
Add sample tests

MitchellGoodwin committed:
Refer to samples

justinmc reviewed (commented):
I like the current approach with the optional useNestedNavigation. I should do another pass through the docs after the examples are finalized to make sure everything is clear and points users in the right place, but overall this looks good.

MitchellGoodwin committed:
Clean up curves

MitchellGoodwin committed:
Clean up animations

MitchellGoodwin committed:
dispose temporary curved animation

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,414 @@
+
+    final Animation<BorderRadiusGeometry> radiusAnimation = curvedAnimation.drive(decorationTween);
+
+    SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle.light);
```
No, it should always be light style. Dark mode already has a dark background, so needs the light contrast. In light mode, the background changes to black, so it needs to switch to a light contrast as well.

chunhtai replied:
do we need to set it back at some point?

MitchellGoodwin replied:
No, once the sheet route is popped, and this delegatedTransition is removed, then the original default will be used.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,414 @@
+/// See also:
+///
+///  * [CupertinoSheetRoute], which is a [PageRoute] that leverages this mixin.
+mixin CupertinoSheetRouteTransitionMixin<T> on PageRoute<T> {
```
Honestly, I'm unsure. I was mostly copying how Cupertino and Material routes are built with their route transition mixins.

chunhtai replied:
we should keep this private or just build into the sheet, I can't think of a good use case user will want to mix this in

chunhtai replied:
in other class they are mainly exposed to be used in other widget to behavior differently when it is under a certain route. This is not the case for sheet at least for now.

chunhtai replied:
also if you decide to keep this class, you should group delegateTransition  override to this class as well

MitchellGoodwin replied:
I'll make this private, and keep it separately just so the architecture of the code is similar to the other routes. I'm sure if there's a need for it to be public, somebody will let us know.

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,414 @@
+
+  @override
+  Widget build(BuildContext context) {
+    return ColoredBox(
+      color: CupertinoColors.transparent,
```
Maybe [SizedBox.expand](https://main-api.flutter.dev/flutter/widgets/SizedBox/SizedBox.expand.html)?

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,414 @@
+            return CupertinoPageRoute<void>(
+              builder: (BuildContext context) {
+                return PopScope(
+                  canPop: settings.name != '/',
```
I think what this does now as-written is to prevent programmatically popping the nested Navigator when on route '/'. So I would double check that that is necessary.

MitchellGoodwin replied:
`onGenerateInitialRoutes` makes sense. Is using both `PopScope` and `NavigatorPopHandler`? Trying to cover both system back gestures and programatic pops.

chunhtai replied:
using popScope and NavigatorPopHandler both sounds good, can you add test for the 2 scenarios?

1. calling Navigator.of(context).pop can pop sheet with nestNavigator
2. using system back gesture on sheet with nestnavigator that has multiple sub routes pops each route correctly and finally pops the entire sheet at the end?

chunhtai commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,500 @@
+      reverseCurve: Curves.easeInToLinear,
+      parent: widget.secondaryRouteAnimation
+    );
+     _primaryPositionAnimation = (_primaryPositionCurve ?? widget.primaryRouteAnimation)
```
same for other property below

MitchellGoodwin replied:
The CurvedAnimations are disposed of on widget dispose and update, so this is a failsafe in case there's a spot were it is null and not set. Theoretically that shouldn't happen. I copied the implementation for Cupertino route https://github.com/flutter/flutter/blob/885a5ca265cb74355d3741da249c3006f58a2427/packages/flutter/lib/src/cupertino/route.dart#L528

chunhtai replied:
> his is a failsafe in case there's a spot were it is null and not set.

but we just set it in line 283, I can't think of a code path this can happen

chunhtai commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,500 @@
+  /// that sheet.
+  static void popSheet(BuildContext context) {
+    if (hasParentSheet(context)) {
+      Navigator.of(context, rootNavigator: true).pop();
```
If we provide useRootNavigator in showCupertinoSheet this will also need to be updated

chunhtai commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,500 @@
+  bool get barrierDismissible => false;
+
+  @override
+  String? get barrierLabel => 'Stacked card appearance for modal bottom sheet';
```
just return null.

chunhtai commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,500 @@
+  }
+
+  @override
+  bool updateShouldNotify(_CupertinoSheetScope oldWidget) => oldWidget.key != null && key != null && oldWidget.key != key;
```
but key isn't a parameter in constructor?

chunhtai commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,500 @@
+    required super.child,
+  });
+
+  static _CupertinoSheetScope? maybeOf(BuildContext context) {
```
Also all the call is side is not using this to build a different widget, this should use getInheritedWidgetOfExactType to not create a dependecy.
also updateShouldNotifty should just return false

chunhtai replied:
even if they do use this to build different content, i can think of a situation this will change in the life time of the widget in the subtree because it would only change if the route is disposed and the widget subtree would have been tear down already.

chunhtai reviewed (commented):
The approach looks good to me, I left some comment about the implementation details

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,500 @@
+    };
+  }
+
+  return Navigator.of(context, rootNavigator: true).push<T>(CupertinoSheetRoute<T>(
```
useRootNavigator was removed because we always want to go to the root when adding a new sheet. The sheet transition does not look correct if it tries to animate within a nested navigator.

chunhtai replied:
sgtm

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,414 @@
+/// will be shown with no [Navigator] widget. Multiple calls to `showCupertinoSheet`
+/// can still be made to show multiple stacked sheets, if desired.
+///
+/// In both cases, new [CupertinoSheetRoute]s will be pushed to the root
```
"Both cases" was supposed to refer to whether or not useNestedNavigation was set to true, but that was impossible to understand. I'll clarify.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,414 @@
+import 'colors.dart';
+import 'route.dart';
+
+// Offset from offscreen below to stopping below the top of the screen.
```
I've been sticking with sheet or Cupertino Sheet. "Modal bottom sheet" is more of a Material thing, I think. Just "sheet" is how the iOS widget is referred.

MitchellGoodwin committed:
Clean up documentation, slight refactor

MitchellGoodwin commented on examples/api/lib/cupertino/sheet/cupertino_sheet.3.dart:
```diff
@@ -0,0 +1,174 @@
+import 'package:flutter/cupertino.dart';
+
+/// Flutter code sample for [CupertinoSheetRoute] with restorable state.
+/// Do we include?
```
I don't think showCupertinoSheet can be used with restorable state, because it uses a regular push internally. [showDatePicker](https://api.flutter.dev/flutter/material/showDatePicker.html) works the same way.

MitchellGoodwin committed:
trim samples

MitchellGoodwin committed:
Address comments

MitchellGoodwin committed:
Add tests for back gesture and pop function

MitchellGoodwin committed:
Add route return type

justinmc commented on examples/api/lib/cupertino/sheet/cupertino_sheet.0.dart:
```diff
@@ -0,0 +1,85 @@
+    );
+  }
+}
+
+
```
Too many newlines here.

justinmc commented on examples/api/lib/cupertino/sheet/cupertino_sheet.0.dart:
```diff
@@ -0,0 +1,85 @@
+            CupertinoButton.filled(
+              onPressed: () {
+                Navigator.of(context).push(CupertinoSheetRoute<void>(
+                  builder: (BuildContext context) => const SheetScaffold()
```
Super nit: Make SheetScaffold private so it's clear that this is something defined in the example and not in the framework?

Here and some others below.

justinmc replied:
This looks good now, thanks.

justinmc commented on examples/api/lib/cupertino/sheet/cupertino_sheet.1.dart:
```diff
@@ -0,0 +1,128 @@
+  }
+}
+
+class SheetNextPage extends StatelessWidget {
+  const SheetNextPage({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return const CupertinoPageScaffold(
+      backgroundColor: CupertinoColors.activeOrange,
+      child: SheetBody(title: 'Next Page')
+    );
+  }
+}
```
Super nit: You could also reuse SheetScaffold but take `title` and `backgroundColor` parameters.

justinmc commented on examples/api/lib/cupertino/sheet/cupertino_sheet.2.dart:
```diff
@@ -0,0 +1,297 @@
+  static Route<void> _counterSheetBuilder(BuildContext context, Object? arguments) {
+    return CupertinoSheetRoute<int?>(
+      builder: (BuildContext context)  {
+        return Navigator(
+          restorationScopeId: 'nested-nav',
```
Can you double check that this works for restoring the current nested page? On Android you could set "Don't keep activities" to true in the device's developer settings, or maybe iOS has something similar. Then navigate to a non-root route in this nested Navigator, switch to another app, switch back, and you should still be at the same nested route.

MitchellGoodwin replied:
I tested against iOS on a device, and was able to navigate away, stop the app in XCode, then navigate back and be at the correct route and have the right counter value.

justinmc commented on examples/api/test/cupertino/sheet/cupertino_sheet.1_test.dart:
```diff
@@ -0,0 +1,26 @@
+
+    await tester.tap(find.text('Go Back'));
+    await tester.pumpAndSettle();
+    expect(dialogTitle, findsNothing);
```
Maybe also try navigating to another sheet and/or another route within the sheet?

justinmc replied:
Thanks again 👍

justinmc commented on examples/api/test/cupertino/sheet/cupertino_sheet.2_test.dart:
```diff
@@ -0,0 +1,26 @@
+
+    await tester.tap(find.text('Pop Sheet'));
+    await tester.pumpAndSettle();
+    expect(dialogTitle, findsNothing);
```
Ideally you should also test the state restoration part, like the scenario I mentioned in my comment on this example above. You can simulate restoration in a test with [restartAndRestore](https://main-api.flutter.dev/flutter/flutter_test/WidgetTester/restartAndRestore.html), see an example of where I did that: https://github.com/flutter/flutter/pull/153723/files#diff-114784af50a13ab49960c747162f9856ddc1b9648a902296e8c50d0bc24de7c1

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,499 @@
+  end: const Offset(0.0, -0.02),
+);
+
+// Tween that animates a sheet slightly up when it is covered by a new sheet..
```
Two period typo at the end here.

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,499 @@
+// Amount the sheet in the background scales down. Found by measuring the width
+// of the sheet in the background and comparing against the screen width on the
+// iOS simulator showing an iPhone 16 pro running iOS 16.0. The scale transition
+//will go from a default of 1.0 to 1.0 - _kSheetScaleFactor.
```
Missing a space here.

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,499 @@
+
+final Animatable<double> _kScaleTween = Tween<double>(begin: 1.0, end: 1.0 - _kSheetScaleFactor);
+
+/// Shows a Cupertino style sheet widget that slides up from the bottom of the
```
Nit: Cupertino-style

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,499 @@
+/// See also:
+///
+///  * [CupertinoSheetRoute] the basic route version of the sheet view.
+///  * [showCupertinoDialog] which displays an iOS-styled dialog.
```
Should we link to the Swift UI docs, or is there anywhere else in Apple's docs that describes what a sheet is? https://developer.apple.com/documentation/swiftui/view/sheet(ispresented:ondismiss:content:)

MitchellGoodwin replied:
HIGs has a page on sheets, so I'll link to there.

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,499 @@
+/// iOS sheet widgets are generally designed to be tightly coupled to the context
+/// of the widget that opened the sheet. As such, it is not recommended to push
+/// a non-sheet route that covers the sheet without first popping the sheet. If
+/// necessary however, it can be done by pushing to the root Navigator.
```
Use square brackets for [Navigator].

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,499 @@
+  late Animation<Offset> _primaryPositionAnimation;
+  // When this page is coming in to cover another sheet.
+  late Animation<Offset> _primaryPositionAnimationWhenCoveringOtherSheet;
+  // The offset animation when this page is becoming covered by another sheet.
```
Nit maybe if this sounds better: "becoming covered" => "being covered"?

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,499 @@
+  }
+
+  // Slightly darkens the sheet behind. Eyeballed from a simulator running iOS 18.0
+  // TODO(mitchgoodwin): Adjust with darkmode logic.
```
Reminder not to miss this TODO.

justinmc commented on packages/flutter/test/cupertino/sheet_test.dart:
```diff
@@ -0,0 +1,547 @@
+import '../widgets/navigator_utils.dart';
+
+void main() {
+
```
Super nit: I think typically no whitespace here.

justinmc commented on packages/flutter/test/cupertino/sheet_test.dart:
```diff
@@ -0,0 +1,547 @@
+      return Column(
+        children: <Widget>[
+          const Text('Page 2'),
+          CupertinoButton(
+          onPressed: () {
```
Some weird indentation here.

justinmc reviewed (commented):
I think the current 3 examples are great and cover what we want to cover in a minimal way.

Everything looks good minus some pretty minor comments here. I think maybe testing state restoration in cupertino_sheet.2.dart is the biggest thing missing, besides the drag to dismiss gesture.

MitchellGoodwin committed:
Add example tests and formatting

MitchellGoodwin committed:
Formatting

justinmc reviewed (commented):
In my [drag review](https://github.com/flutter/flutter/pull/161324#pullrequestreview-2540646328) I left a comment about it not looking right in dark mode. I think it looks a little better here if I'm not mistaken, but still not right?

<details>

<summary>Dark mode screenshot, first example</summary>

<img width="476" alt="Screenshot 2025-01-09 at 11 46 45 AM" src="https://github.com/user-attachments/assets/a31e047c-a485-44fd-9488-6115eb5fb98e" />

</details>

This PR is just about ready to land I think. The remaining things I see:

  * Drag. Address the comments in [my drag review](https://github.com/flutter/flutter/pull/161324#pullrequestreview-2540646328), or create issues for things that will be fixed later. Merge/move the drag commit to this PR.
  * Decide what to do about dark mode.

I think that's it and then I'm happy to LGTM.

MitchellGoodwin committed:
Added overlay for dark mode

MitchellGoodwin commented:
I added an overlay color that gets stacked over the previous sheet to make it stand out in dark mode. I also added elevation to the sheet itself which caused the background color to match Swift's sheet.

Left is Flutter, right is native.
<img width="841" alt="Screenshot 2025-01-09 at 3 24 14 PM" src="https://github.com/user-attachments/assets/fc348d71-fee0-4250-b6ed-15b6745df9bd" />

MitchellGoodwin committed:
Add drag dismiss gesture to sheet

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,541 @@
+
+  /// Pops the entire [CupertinoSheetRoute], if a sheet route exists in the stack.
+  ///
+  /// Used if to pop an entire sheet at once, if there is nested naviagtion within
```
naviagtion => navigation

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,541 @@
+///
+/// Typically called by [showCupertinoSheet], which provides some boilerplate for
+/// pushing the `CupertinoSheetRoute` to the root navigator and providing simple
+/// nested naviagation.
```
naviagation => navigation

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,541 @@
+            ? Stack(
+              children: <Widget>[
+                child,
+                FadeTransition(
```
Thanks for getting the details right and even animating dark mode 👍

justinmc reviewed (commented):
LGTM 👍 This is assuming you'll cherry pick in the drag commit (https://github.com/flutter/flutter/pull/161324/commits/73d4dcffd2b52721214cd197768344ca6095df48 or a modified later one) and there are no significant conflicts.

Thanks for doing dark mode, it looks perfect by my eye. I think that was the biggest weakness of this PR. The only remaining imperfections are much more subtle. I think people will be happy.

Make sure you consider this comment if you missed it (though it's a nit): https://github.com/flutter/flutter/pull/157568/files#r1909358363

MitchellGoodwin committed:
Addressing nits

MitchellGoodwin committed:
Addressing nits

justinmc commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,767 @@
+    _downGestureController!.dragUpdate(
+      // Devide by size of the sheet. The gap between the top of the sheet and
+      // top of the screen is 0.08.
+      details.primaryDelta! / (context.size!.height - (context.size!.height * 0.08)),
```
Nit comment from the other PR about this 0.08 value: https://github.com/flutter/flutter/pull/161324/files#r1911057792

MitchellGoodwin committed:
Get screen gap value from constant

justinmc reviewed (approved):
Official LGTM 👍 ™️

Did you look at the post frame callback thing? Sorry if you responded somewhere else and I missed it. https://github.com/flutter/flutter/pull/161324#discussion_r1909315735

MitchellGoodwin commented:
> Did you look at the post frame callback thing? Sorry if you responded somewhere else and I missed it.

The drag gesture for normal Cupertino pages had that logic from this PR: https://github.com/flutter/flutter/pull/139585.
Seems it was handling a go_router bug. I assumed we'd need it as well to avoid that bug. I'm not 100% sure why it needs to happen a frame later, but I assume it either keeps the transitions from stuttering (transitions change depending on if a user gesture is in progress or not) or it just logically reports that the gesture has ended after the controller disposal when theoretically the user is still touching the area, not simultaneously.

chunhtai commented:
> > Did you look at the post frame callback thing? Sorry if you responded somewhere else and I missed it.
> 
> The drag gesture for normal Cupertino pages had that logic from this PR: #139585. Seems it was handling a go_router bug. I assumed we'd need it as well to avoid that bug. I'm not 100% sure why it needs to happen a frame later, but I assume it either keeps the transitions from stuttering (transitions change depending on if a user gesture is in progress or not) or it just logically reports that the gesture has ended after the controller disposal when theoretically the user is still touching the area, not simultaneously.

I forgot why the shcedule postframe is needed, but i remember there was a reason, cc @hannah-hyj do you still remember why we need post frame callback instead of renable the gesture directly?

chunhtai commented on packages/flutter/lib/src/cupertino/sheet.dart:
```diff
@@ -0,0 +1,768 @@
+}
+
+class _CupertinoSheetTransitionState extends State<CupertinoSheetTransition> {
+  // The offset animation when this page is being covered by another sheet.
```
Can you add some newline to make it more readable

chunhtai reviewed (approved):
LGTM

MitchellGoodwin committed:
Add new lines

justinmc commented:
It sounds like it would be good to have a comment explaining the post frame callback if possible, but I'm happy either way.

MaherSafadii commented:
> I added an overlay color that gets stacked over the previous sheet to make it stand out in dark mode. I also added elevation to the sheet itself which caused the background color to match Swift's sheet.
> 
> Left is Flutter, right is native. <img alt="Screenshot 2025-01-09 at 3 24 14 PM" width="841" src="https://private-user-images.githubusercontent.com/58190796/401730335-fc348d71-fee0-4250-b6ed-15b6745df9bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY1NDQ0MTMsIm5iZiI6MTczNjU0NDExMywicGF0aCI6Ii81ODE5MDc5Ni80MDE3MzAzMzUtZmMzNDhkNzEtZmVlMC00MjUwLWI2ZWQtMTViNjc0NWRmOWJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEwVDIxMjE1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4OWYwZDMzMjY1MjBlY2JmODMzZGRmMDg5MGE2YjYzNWJiMjNlMTE1MGQwODYzNjRkNGM5ZWE3NTFmMDI0NGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JWdP04mWXVZbqxymnUFYsXvWh-epVKo5mPKian2LqxA">

The overlay works but its only for the page that's being presented and not on the one in the back too, if background color of a page is black and you push a sheet it won't be visible behind the sheet which isn't the correct behaviour like on the native side, here's a video that shows what I mean: 

https://github.com/user-attachments/assets/81ae2eb7-2a8a-4bf2-9227-03160d94e7d5

MaherSafadii commented:
> > I added an overlay color that gets stacked over the previous sheet to make it stand out in dark mode. I also added elevation to the sheet itself which caused the background color to match Swift's sheet.
> > Left is Flutter, right is native. <img alt="Screenshot 2025-01-09 at 3 24 14 PM" width="841" src="https://private-user-images.githubusercontent.com/58190796/401730335-fc348d71-fee0-4250-b6ed-15b6745df9bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY1NDQ0MTMsIm5iZiI6MTczNjU0NDExMywicGF0aCI6Ii81ODE5MDc5Ni80MDE3MzAzMzUtZmMzNDhkNzEtZmVlMC00MjUwLWI2ZWQtMTViNjc0NWRmOWJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEwVDIxMjE1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4OWYwZDMzMjY1MjBlY2JmODMzZGRmMDg5MGE2YjYzNWJiMjNlMTE1MGQwODYzNjRkNGM5ZWE3NTFmMDI0NGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JWdP04mWXVZbqxymnUFYsXvWh-epVKo5mPKian2LqxA">
> 
> The overlay works but its only for the page that's being presented and not on the one in the back too, if background color of a page is black and you push a sheet it won't be visible behind the sheet which isn't the correct behaviour like on the native side, here's a video that shows what I mean:
> 
>  Recording.at.2025-01-10.23.35.47.mp4

turns out that the PR has code to make this functionality work but its checking the platform brightness instead of the apps brightness, so the device has to be in dark mode for it to behave correctly, so if the app is in dark mode but the device is in light mode, it won't work properly.

This line is the one is the PR used to check the theme
`    final bool isDarkMode = MediaQuery.of(context).platformBrightness == Brightness.dark;
`

I switched it with this line which checks the apps brightness instead of the device's and it worked correctly:
`final bool isDarkMode = CupertinoTheme.brightnessOf(context) == Brightness.dark;`

MitchellGoodwin committed:
Check theme brightness instead of platform

MitchellGoodwin commented:
> turns out that the PR has code to make this functionality work but its checking the platform brightness instead of the apps brightness, so the device has to be in dark mode for it to behave correctly, so if the app is in dark mode but the device is in light mode, it won't work properly.

Good catch.

hannah-hyj commented:
> > > Did you look at the post frame callback thing? Sorry if you responded somewhere else and I missed it.
> > 
> > 
> > The drag gesture for normal Cupertino pages had that logic from this PR: #139585. Seems it was handling a go_router bug. I assumed we'd need it as well to avoid that bug. I'm not 100% sure why it needs to happen a frame later, but I assume it either keeps the transitions from stuttering (transitions change depending on if a user gesture is in progress or not) or it just logically reports that the gesture has ended after the controller disposal when theoretically the user is still touching the area, not simultaneously.
> 
> I forgot why the shcedule postframe is needed, but i remember there was a reason, cc @hannah-hyj do you still remember why we need post frame callback instead of renable the gesture directly?

I tested it again, look like the schedule postframe in #139585 is needed because changing user gesture status in dispose() will cause the rebuild error "markNeedsBuild() called when widget tree was locked", adding a post frame will avoid that.

JavierPerezLavadie commented:
@MitchellGoodwin Wow amazing job

definev commented:
How to integrate `CupertinoSheetRoute` with go_router? I wrapped it to `Page` but it throw `'package:flutter/src/widgets/navigator.dart': Failed assertion: line 3110 pos 15: '!pageBased || route.settings is Page': is not true.`
```dart
class CupertinoSheetPage extends Page<void> {
  const CupertinoSheetPage({
    required this.child,
    super.key,
  });

  final Widget child;

  @override
  Route<void> createRoute(BuildContext context) {
    return CupertinoSheetRoute(
      builder: (context) => child,
    );
  }
}
```

MitchellGoodwin commented:
> When is nested scrolling going to be added? I have tried all scrolling hacks. In the provided picture, I should be able to scroll to a third color.

Created an issue: #161623. It's a quick fix to enable nested scrolling. Getting the nested scrolling to smoothly interact with the drag dismiss gesture will be a longer effort.

> Also-- I had to create my own appbar to exist in a stack above the modal sheet. 

We don't currently have anything on the backlog for adding a sheet specific app bar. Could you create an issue?

In general we're looking for issues for improvements to be made to the sheet.

MaherSafadii commented:
> > When is nested scrolling going to be added? I have tried all scrolling hacks. In the provided picture, I should be able to scroll to a third color.
> 
> Created an issue: #161623. It's a quick fix to enable nested scrolling. Getting the nested scrolling to smoothly interact with the drag dismiss gesture will be a longer effort.
> 
> > Also-- I had to create my own appbar to exist in a stack above the modal sheet.
> 
> We don't currently have anything on the backlog for adding a sheet specific app bar. Could you create an issue?
> 
> In general we're looking for issues for improvements to be made to the sheet.

I dont think new appbar is needed, the current CupertinoNavigationBar & CupertinoSliverNavigationBar are fine, all that's needed is change to their code to check if they're in a sheets context, if yes, change padding values and other small tweaks that make it match native iOS.

MaherSafadii commented:
> > > When is nested scrolling going to be added? I have tried all scrolling hacks. In the provided picture, I should be able to scroll to a third color.
> > 
> > 
> > Created an issue: #161623. It's a quick fix to enable nested scrolling. Getting the nested scrolling to smoothly interact with the drag dismiss gesture will be a longer effort.
> > > Also-- I had to create my own appbar to exist in a stack above the modal sheet.
> > 
> > 
> > We don't currently have anything on the backlog for adding a sheet specific app bar. Could you create an issue?
> > In general we're looking for issues for improvements to be made to the sheet.
> 
> I dont think new appbar is needed, the current CupertinoNavigationBar & CupertinoSliverNavigationBar are fine, all that's needed is change to their code to check if they're in a sheets context, if yes, change padding values and other small tweaks that make it match native iOS.

for example, I tried modifying Flutter's CupertinoNavigationBar to check if the current context in in a sheet or not, if yes it should remove the top SaveArea padding, here are pics of the before and after, also the transition on push glitches the nav bar, you have to explicitly set the set transitionBetweenRoutes to false in the sheets nav bar, it should  automatically tell if its in a sheet and disable it.

![Uploading Screenshot 2025-01-15 at 1.48.51 PM.png…]()
<img width="1512" alt="Screenshot 2025-01-15 at 1 49 05 PM" src="https://github.com/user-attachments/assets/ac457088-a4ae-4665-84d1-5a3feb6e60b5" />

masal9pse commented:
From which version of Flutter will CupertinoSheetRoute be available?

MaherSafadii commented:
> From which version of Flutter will CupertinoSheetRoute be available?

The next stable version I think, you can try it now by switching to the main channel, if you want to try it without switching channels, you can copy the sheet code from this PR and directly add it to your flutter project which I did and works fine.

MaherSafadii commented:
> From which version of Flutter will CupertinoSheetRoute be available?

The next stable version I think, you can try it now by switching to the main channel, if you want to try it without switching channels, you can copy the sheet code from this PR and directly add it to your flutter project which I did and works fine.

MasatoYamamotoo commented:
@MaherSafadii 
https://github.com/flutter/flutter/pull/157568#issuecomment-2592535332

Regarding this one, I am also seeing unnecessary padding in iOS, how can I set NavigationToolbar to reflect that in appBar?
Would it be possible for you to provide a minimum sample code?
